File: 許靜玟/2022.02.11/SpiralMatrixII.cpp
class Solution {
public:
    vector<vector<int> > generateMatrix(int n) {
        vector<vector<int> > ans(n, vector<int>(n));
        int row= 0;
        int col= 0;
        int rows= n-1;
        int cols= n-1;
        int num = 1;
        int total =n*n;
        while(num <= total)
        {
            for(int index= col; index<= cols; index++) //top
            {
                ans[row][index] = num++;
            }
            row++;
            for(int index= row; index<= rows; index++) //right
            {
                ans[index][cols] = num++;
            }
            cols--;
            for(int index= cols; index>= col; index--) //button
            {
                ans[rows][index] = num++;
            }
            rows--;
            for(int index= cols; index>= row; index--) //left
            {
                ans[index][col] = num++;
            }
            col++;
        }
        return ans;
    }
};

File: 陳柏諺/2.11/Spiral Matrix II.js
/**
 * 1. 依題目要求左到右，上到下
 * @param {number} n
 * @return {number[][]}
 */
const generateMatrix = function(n) {
    let ans = Array(n).fill(0).map(n => Array(n));
    let left = 0;
    let right = n - 1;
    let top = 0;
    let bottom = n - 1;
    let num = 1;
    while (left < right) {
        // left->right
        for (let i = left; i < right; i++) {
            ans[top][i] = num++;
        }
        // top->bottom
        for (let i = top; i < bottom; i++) {
            ans[i][right] = num++;
        }
        // right->left
        for (let i = right; i > left; i--) {
            ans[bottom][i] = num++;
        }
        // bottom->top
        for (let i = right; i > left; i--) {
            ans[i][left] = num++;
        }
        left++;
        right--;
        top++;
        bottom--;
    }
    if (n % 2 == 1) {
        n = Math.floor(n / 2);
        ans[n][n] = num;
    }
    return ans
};